server/config/server.ts

```diff
@@ -0,0 +1,206 @@
+import path from 'path'
+import fs from 'fs'
+import yaml from 'yaml'
+import deepMerge from 'deepmerge'
+import { PartialDeep } from 'type-fest'
+import { ACL } from '../util/restrict'
+
+export type ProviderConfig = {
+  name: string;
+  options: unknown;
+}
+
+export type Sponsor = {
+  name: string;
+  icon: string;
+  description: string;
+  small?: boolean;
+}
+
+export type ServerConfig = {
+  database: {
+    sql: string | {
+      host: string;
+      port: number;
+      user: string;
+      password: string;
+      database: string;
+    }
+    redis: string | {
+      host: string;
+      port: number;
+      password: string;
+      database: number;
+    }
+    migrate: string;
+  }
+  instanceType: string;
+  tokenKey: string;
+  origin: string;
+
+  ctftime?: {
+    clientId: string;
+    clientSecret: string;
+  }
+
+  userMembers: boolean;
+  sponsors: Sponsor[];
+  homeContent: string;
+  ctfName: string;
+  meta: {
+    description: string;
+    imageUrl: string;
+  }
+  logoUrl?: string;
+  globalSiteTag?: string;
+
+  challengeProvider: ProviderConfig;
+  uploadProvider: ProviderConfig;
+
+  email?: {
+    provider: ProviderConfig;
+    from: string;
+  }
+
+  divisions: Record<string, string>;
+  defaultDivision?: string;
+  divisionACLs?: ACL[];
+
+  startTime: number;
+  endTime: number;
+
+  leaderboard: {
+    maxLimit: number;
+    maxOffset: number;
+    updateInterval: number;
+    graphMaxTeams: number;
+    graphSampleTime: number;
+  }
+  loginTimeout: number;
+}
+
+const configPath = path.join(__dirname, '../../conf.d')
+const files = fs.readdirSync(configPath).filter(name => /\.(?:ya?ml|json)$/.test(name)).sort()
+const fileConfigs = files.map(name => yaml.parse(fs.readFileSync(path.join(configPath, name)).toString()))
+
+const parseBoolEnv = (val?: string): boolean | null => {
+  if (val == null) {
+    return null
+  }
+  return ['true', 'yes', 'y'].includes(val.toLowerCase().trim())
+}
+
+const envConfig: ServerConfig = {
```

Comment:
This should be defined as a deep partial; otherwise this will fail strict null checks.